### PR TITLE
Omnibus s3 caching

### PR DESCRIPTION
**Pull request by gmmeyer**
### What does this PR do?

This PR adds s3 caching to Omnibus. It requires this PR to be merged in order to use it: https://github.com/DataDog/omnibus-ruby/pull/52

### Motivation

This should speed up builds significantly and prevent sourceforge and others from mucking it up!

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -32,8 +32,10 @@ variables:
   S3_CP_OPTIONS: --only-show-errors --region us-east-1 --sse AES256
   S3_CP_CMD: aws s3 cp $S3_CP_OPTIONS
   S3_ARTEFACTS_URI: s3://dd-ci-artefacts-build-stable/$CI_PROJECT_NAME/$CI_PIPELINE_ID
+  S3_OMNIBUS_CACHE_BUCKET: dd-ci-datadog-agent-omnibus-cache-build-stable
   S3_DSD6_URI: s3://dsd6-staging/linux
   RELEASE_VERSION: nightly
+  BASE_ROLE_SESSION_NAME: $CI_PROJECT_NAME-$CI_PIPELINE_ID
```
Comment (olivielpeau) [inline .gitlab-ci.yml]: what's this var for? Don't see it used anywhere

Let's document what it does if it's useful :)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -152,16 +155,16 @@ build_dogstatsd-deb_x64:
...
+# run_dogstatsd_size_test:
+#   stage: integration_test
+#   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/deb_x64:latest
+#   tags: [ "runner:main", "size:large" ]
+#   before_script:
+#     # Disable global before_script
+#     - mkdir -p $STATIC_BINARIES_DIR
+#     - $S3_CP_CMD $S3_ARTEFACTS_URI/static/dogstatsd $STATIC_BINARIES_DIR/dogstatsd
+#   script:
+#     - inv -e dogstatsd.size-test --skip-build
```
Comment (olivielpeau) [inline .gitlab-ci.yml]: let's uncomment this block :)

File: omnibus/Gemfile [hunk 3]
```diff
@@ -1,7 +1,7 @@
 source 'https://rubygems.org'
 
 # Install omnibus
-gem 'omnibus', git: 'https://github.com/DataDog/omnibus-ruby.git', branch: 'datadog-5.5.0'
+gem 'omnibus', git: 'https://github.com/DataDog/omnibus-ruby.git', branch: 'greg/patch'
```
Comment (olivielpeau) [inline omnibus/Gemfile]: let's undo this

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -173,9 +176,12 @@ agent_deb-x64:
   stage: package_build
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/deb_x64:latest
   tags: [ "runner:main", "size:large" ]
+  variables:
+    AWS_CONTAINER_CREDENTIALS_RELATIVE_URI: /credentials
   script:
     # remove artifacts from previous pipelines that may come from the cache
     - rm -rf $OMNIBUS_PACKAGE_DIR/*
+    - echo $ROLE_SESSION_NAME
```
Comment (olivielpeau) [inline .gitlab-ci.yml]: let's remove this? (unless it's indeed useful)

File: tasks/agent.py [hunk 5]
```diff
@@ -234,13 +234,15 @@ def omnibus_build(ctx, puppy=False, log_level="info", base_dir=None, gem_path=No
...
+        cmd = "{omnibus} build {project_name} --log-level={log_level} {populate_s3_cache} {overrides}"
         args = {
             "omnibus": omnibus,
             "project_name": "puppy" if puppy else "agent",
             "log_level": log_level,
-            "overrides": overrides_cmd
+            "overrides": overrides_cmd,
+            "populate_s3_cache": ""
         }
+        args['populate_s3_cache'] = " --populate-s3-cache "
```
Comment (olivielpeau) [inline tasks/agent.py]: let's set this value directly in the `args` dict when it's define above

**Review by olivielpeau — COMMENTED** [review]
LGTM overall, just a few nits!

File: .gitlab-ci.yml (hunk 2 above)
Comment (gmmeyer) [reply .gitlab-ci.yml]: good spot thank you!

File: omnibus/Gemfile (hunk 3 above)
Comment (gmmeyer) [reply omnibus/Gemfile]: I'll merge the other PR and then remove it

**Comment by gmmeyer** [comment]
mind approving now? :)

File: tasks/dogstatsd.py [hunk 6]
```diff
@@ -170,12 +170,14 @@ def omnibus_build(ctx, log_level="info", base_dir=None, gem_path=None,
...
-        cmd = "{omnibus} build dogstatsd --log-level={log_level} {overrides}"
+        cmd = "{omnibus} build dogstatsd --log-level={log_level} {populate_s3_cache} {overrides}"
         args = {
             "omnibus": omnibus,
             "log_level": log_level,
-            "overrides": overrides_cmd
+            "overrides": overrides_cmd,
+            "populate_s3_cache": ""
         }
+        args['populate_s3_cache'] = " --populate-s3-cache "
```
Comment (olivielpeau) [inline tasks/dogstatsd.py]: same comment here as for `tasks/agent.py`

**Review by olivielpeau — COMMENTED** [review]
one last nit, other than that LGTM!

I can take care of merging this right after the end of the 6.2 freeze. Want to avoid breaking anything right before the 6.2 freeze :)

File: tasks/agent.py [hunk 7]
```diff
@@ -235,12 +235,13 @@ def omnibus_build(ctx, puppy=False, log_level="info", base_dir=None, gem_path=No
...
         omnibus = "bundle exec omnibus.bat" if invoke.platform.WINDOWS else "bundle exec omnibus"
-        cmd = "{omnibus} build {project_name} --log-level={log_level} {overrides}"
+        cmd = "{omnibus} build {project_name} --log-level={log_level} {populate_s3_cache} {overrides}"
         args = {
             "omnibus": omnibus,
             "project_name": "puppy" if puppy else "agent",
             "log_level": log_level,
-            "overrides": overrides_cmd
+            "overrides": overrides_cmd,
+            "populate_s3_cache": " --populate-s3-cache "
```
Comment (olivielpeau) [inline tasks/agent.py]: Actually, using this flag makes the build fail when the build environment doesn't have permissions to push to S3. So let's make this a flag of the invoke command, and enable the flag in the gitlab build only.

File: tasks/dogstatsd.py (hunk 6 above)
Comment (olivielpeau) [reply tasks/dogstatsd.py]: (same comment as `tasks/agent.py`, let's make this a flag)

**Review by olivielpeau — CHANGES_REQUESTED** [review]
Realized this would break the omnibus command on environments that don't have access to S3, @gmmeyer could you address my comment?

File: tasks/agent.py (hunk 7 above)
Comment (gmmeyer) [reply tasks/agent.py]: sounds like a good idea! I will fix this!

**Comment by gmmeyer** [comment]
@olivielpeau I put it behind a flag

**Review by olivielpeau — APPROVED** [review]
LGTM! 🎉 

I've fixed a small typo, merging this

**Pull request merged**